### PR TITLE
Collijk/feature/mandate reimposition

### PR DESCRIPTION
**Pull request by collijk**
This PR implements the mandate reimposition algorithm, removing the external loop from the production pipeline.  

This is very much a "make it work" PR, which it does.  It needs a pass to turn stuff into functions, build abstractions, and allow a more generic algorithm specification for the forecast, but postprocessing updates come first.

### CODE REVIEW

File: src/covid_model_seiir_pipeline/forecasting/data.py [hunk 1]
```diff
@@ -64,6 +64,15 @@ def load_location_ids(self) -> List[int]:
             location_ids = yaml.full_load(location_file)
         return location_ids
 
+    def load_thetas(self, theta_specification: Union[str, int]) -> pd.Series:
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/data.py:67]: just a lift an move

File: src/covid_model_seiir_pipeline/forecasting/data.py [hunk 2]
```diff
@@ -111,11 +120,12 @@ def load_regression_coefficients(self, draw_id: int) -> pd.DataFrame:
         return self.regression_marshall.load(MKeys.coefficient(draw_id))
 
     # TODO: inverse is RegressionDataInterface.save_date_file
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/data.py:122]: update comment

File: src/covid_model_seiir_pipeline/forecasting/data.py [hunk 3]
```diff
@@ -111,11 +120,12 @@ def load_regression_coefficients(self, draw_id: int) -> pd.DataFrame:
...
-    def load_dates_df(self, draw_id: int) -> pd.DataFrame:
+    def load_transition_date(self, draw_id: int) -> pd.Series:
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/data.py:123]: This is the thing we want anyway.

File: src/covid_model_seiir_pipeline/forecasting/model.py [hunk 4]
```diff
@@ -83,32 +85,96 @@ def __post_init__(self):
...
 
-    def __init__(self, model_specs: SeiirModelSpecs, init_cond: np.ndarray):
-        self.model_specs = model_specs
-        self.init_cond = init_cond
-
-    def get_solution(self, times, beta, theta=None, solver="RK4"):
-        model = CustomizedSEIIR(**asdict(self.model_specs))
-        if solver == "RK4":
-            solver = RK4(model.system, self.model_specs.delta)
+    def __init__(self, solver_name: str, model_specs: SeiirModelSpecs):
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/model.py:88]: Tried a refactor to handle multiple algorithms.  Threw it away after it didn't work and just reoriented the args a bit.  Nothing fancy happening here.

File: src/covid_model_seiir_pipeline/forecasting/model.py [hunk 5]
```diff
@@ -83,32 +85,96 @@ def __post_init__(self):
...
-                                 beta.reshape((1, -1)), theta.reshape((1, -1))], axis=0).T,
-            columns=["S", "E", "I1", "I2", "R", "t", "beta", "theta"]
+            data=result_array,
+            columns=self.model.components + self.model.params + ['t']
         )
+
         return result
 
 
+def run_normal_ode_model_by_location(initial_condition, beta_params, betas, thetas, location_ids, solver):
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/model.py:118]: lift and move of central loop.

File: src/covid_model_seiir_pipeline/forecasting/model.py [hunk 6]
```diff
@@ -83,32 +85,96 @@ def __post_init__(self):
...
+        loc_thetas = np.repeat(thetas.get(location_id, default=0), loc_betas.size)
+
+        forecasted_components = ode_runner.get_solution(init_cond, loc_times, loc_betas, loc_thetas)
+        forecasted_components['date'] = loc_days.values
+        forecasted_components['location_id'] = location_id
+        forecasts.append(forecasted_components)
+    forecasts = pd.concat(forecasts)
+    return forecasts
+
+
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/model.py:146]: Next three functions also just lift and move from the task script.

File: src/covid_model_seiir_pipeline/forecasting/task/beta_forecast.py [hunk 7]
```diff
@@ -26,99 +26,118 @@ def run_beta_forecast(draw_id: int, forecast_version: str, scenario_name: str):
...
-    infections = observed_infections.combine_first(modeled_infections)['cases_draw'].rename('infections')
+    # Modeling starts
+    betas = model.forecast_beta(covariate_pred, coefficients, beta_scales)
+    future_components = model.run_normal_ode_model_by_location(initial_condition, beta_params, betas, thetas,
+                                                               location_ids, scenario_spec.solver)
+    components = model.splice_components(past_components, future_components, transition_day)
+    components['theta'] = thetas.reindex(components.index).fillna(0)
+    infections, deaths, r_effective = model.compute_output_metrics(infection_data, components, beta_params)
+
+    if scenario_spec.algorithm == 'mandate_reimposition':
```
Comment (collijk) [inline src/covid_model_seiir_pipeline/forecasting/task/beta_forecast.py:71]: TODO: turn into functions, add wait and days on to algorithm parameters

File: src/covid_model_seiir_pipeline/forecasting/task/beta_forecast.py [hunk 8]
```diff
@@ -26,99 +26,118 @@ def run_beta_forecast(draw_id: int, forecast_version: str, scenario_name: str):
...
+        min_wait = pd.Timedelta(days=7)
+        days_on = pd.Timedelta(days=7 * 6)
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/forecasting/task/beta_forecast.py:73]: Should this be `min_wait * 6` or is it just coincidence they're both in units of 1 week?

Comment (collijk) [reply src/covid_model_seiir_pipeline/forecasting/task/beta_forecast.py:73]: coincidence

**Pull request merged**